src/StateService/Verification/VerificationService.cs

```diff
@@ -0,0 +1,161 @@
+    {
+        public class ValidatedRootPersisted { public uint Index; }
+        public class BlockPersisted { public uint Index; }
+        private class Timer { public uint Index; }
+        private static readonly uint TimeoutMilliseconds = ProtocolSettings.Default.MillisecondsPerBlock;
+        private static readonly uint DelayMilliseconds = 3000;
+        private readonly NeoSystem core;
+        private const int MaxCachedVerificationProcessCount = 10;
+        private readonly Wallet wallet;
+        private readonly ConcurrentDictionary<uint, VerificationContext> contexts = new ConcurrentDictionary<uint, VerificationContext>();
+
+        public VerificationService(NeoSystem core, Wallet wallet)
+        {
+            this.core = core;
+            this.wallet = wallet;
+        }
+
+        private void SendVote(VerificationContext context)
+        {
+            var vote = context.CreateVote();
+            if (vote is null) return;
+            Utility.Log(nameof(VerificationService), LogLevel.Info, $"relay vote, height={vote.RootIndex}, retry={context.Retries}");
+            Settings.Default.VerifierUrls.ForEach(url =>
+            {
+                Task.Run(() =>
+                {
+                    try
+                    {
+                        var http_client = new HttpClient
+                        {
+                            Timeout = TimeSpan.FromMilliseconds(TimeoutMilliseconds),
+                        };
+                        var client = new RpcClient(http_client, url);
+                        client.RpcSendAsync("votestateroot", vote.RootIndex, vote.ValidatorIndex, Convert.ToBase64String(vote.Signature))
+                            .GetAwaiter().GetResult();
+                    }
+                    catch (Exception e)
+                    {
+                        Utility.Log(nameof(VerificationService), LogLevel.Warning, $"Failed to send vote, verifier={url}, error={e.Message}");
+                    }
+                });
+            });
+        }
+
+        private void OnVoteStateRoot(Vote vote)
+        {
+            if (contexts.TryGetValue(vote.RootIndex, out VerificationContext context) && context.AddSignature(vote.ValidatorIndex, vote.Signature))
+            {
+                CheckVotes(context);
+            }
+        }
+
+        private void CheckVotes(VerificationContext context)
+        {
+            if (context.CheckSignatures())
+            {
+                if (context.Message is null) return;
+                Utility.Log(nameof(VerificationService), LogLevel.Info, $"relay state root, height={context.StateRoot.Index}, root={context.StateRoot.RootHash}");
+                core.Blockchain.Tell(context.Message);
+            }
+        }
+
+        private void OnBlockPersisted(uint index)
+        {
+            if (MaxCachedVerificationProcessCount <= contexts.Count)
+            {
+                var indexes = contexts.Keys.OrderBy(i => i).ToArray();
```

Comment:
Use list?

Reply:
I think `ReadOnlySpan<T>` is better.

Reply:
Done.

Reply:
> I think `ReadOnlySpan<T>` is better.

This `var indexes = new ReadOnlySpan<uint>(contexts.Keys.OrderBy(i => i).ToArray());`?

Reply:
Done.

Reply:
```C#
            if (MaxCachedVerificationProcessCount <= contexts.Count)
            {
                contexts.Keys.OrderBy(p => p).Take(contexts.Count - MaxCachedVerificationProcessCount + 1).ForEach(p =>
                {
                    if (contexts.TryRemove(p, out var value))
                    {
                        value.Timer.CancelIfNotNull();
                    }
                });
            }
```
@erikzhang  How about this？

Reply:
Good.

Reply:
Done.